src/hairy_probes.F

```diff
@@ -0,0 +1,614 @@
+      kTS = 0.0_dp
+
+      !***HP loop
+      DO ispin = 1, nspin
+         DO ikp = 1, nkp
+            CALL HP_occupancy(probe,coeff_squared(:,:,ikp,ispin),energies(:,ikp,ispin), &
+                              maxocc,fermi,occ(:,ikp,ispin),kTS_kp)
+
+            kTS = kTS + kTS_kp * wk(ikp) !entropic contribution
+            N_now = N_now + accurate_sum(occ(1:nMo,ikp,ispin))*wk(ikp) 
+         END DO
+      END DO 
+      !***HP loop
+      
+      DEALLOCATE(coeff_squared)
+
+      CALL timestop(handle)
+END SUBROUTINE probe_occupancy_kp 
+
+!**************************************************************************************
+!
+!
+!
+!**************************************************************************************
+SUBROUTINE HP_occupancy(probe, matrix, energies, maxocc, fermi, occupancy, kTS)
+        IMPLICIT NONE 
+
+        REAL(KIND=dp), INTENT(OUT)                 :: occupancy(:), kTS
+        TYPE(hairy_probes_type), INTENT(IN)         :: probe(:)
+        REAL(KIND=dp), INTENT(IN)                  :: matrix(:,:) !squared coefficient matrix
+        REAL(KIND=dp), INTENT(IN)                  :: energies(:), fermi, maxocc
+
+        INTEGER :: iMO, ip, np, nMO
+        REAL(KIND=dp) :: mu, fermi_fun, fermi_fun_sol, sum_fermi_fun, sum_coeff, &
+                         & sum_fermi_fun_sol, sum_coeff_sol, C, s, f
+
+        nMO = SIZE(matrix, 2)
+        np = SIZE(probe)
+
+        kTS = 0.0_dp
+        ! kTS is the entropic contribution to the electronic energy
+
+        mos2: do iMO = 1, nMO
+
+           sum_fermi_fun = 0.0_dp
+           sum_coeff = 0.0_dp 
+
+           sum_fermi_fun_sol = 0.0_dp
+           sum_coeff_sol = 0.0_dp 
+
+           probes2: DO ip = 1, np
+              IF ( probe(ip)%alpha .lt. 1.0_dp ) THEN 
+                 !fermi distribution, solution probes 
+                 CALL fermi_distribution ( energies(iMO), fermi, probe(ip)%T, fermi_fun_sol )                        
+                 !sum of coefficients, solution probes 
+                 sum_coeff_sol = sum_coeff_sol + SUM(matrix(probe(ip)%first_ao:probe(ip)%last_ao,iMO))
+
+              ELSE  
+                 C = SUM(matrix(probe(ip)%first_ao:probe(ip)%last_ao,iMO))  
+                 !bias probes 
+                 mu = fermi - probe(ip)%mu  
+                 !fermi distribution, main probes                   
+                 CALL fermi_distribution ( energies(iMO), mu, probe(ip)%T, fermi_fun )                        
+                 !sum fermi distribution * coefficients
+                 sum_fermi_fun = sum_fermi_fun + ( fermi_fun * C )
+                 !sum cofficients 
+                 sum_coeff = sum_coeff + C 
+              END IF 
+           END DO probes2
+
+           sum_fermi_fun_sol = probe(np)%alpha * fermi_fun_sol * sum_coeff_sol
+           sum_coeff_sol = probe(np)%alpha * sum_coeff_sol                       
+           f = ( sum_fermi_fun_sol + sum_fermi_fun ) / ( sum_coeff_sol + sum_coeff ) 
```

Comment:
My guess would be that the [Floating-point exception ](https://storage.googleapis.com/cp2k-ci/run-cp2k-sdbg-869dc475_report.txt) are because `sum_coeff_sol + sum_coeff` is sometimes zero.